File: kbe/res/sdk_templates/client/unity/DataTypes.cs
namespace KBEngine
{
  	using UnityEngine; 
	using System; 
	using System.Collections; 
	using System.Collections.Generic;
	
	using MessageID = System.UInt16;
	
	/*
		entitydef所支持的基本数据类型
		改模块中的类抽象出了所有的支持类型并提供了这些类型的数据序列化成二进制数据与反序列化操作（主要用于网络通讯的打包与解包）
	*/
	public class DATATYPE_BASE
	{
		public virtual void bind()
		{
		}
		
		public virtual object createFromStream(MemoryStream stream)
		{
			return null;
		}
		
		public virtual void addToStream(Bundle stream, object v)
		{
		}
		
		public virtual object parseDefaultValStr(string v)
		{
			return null;
		}
		
		public virtual bool isSameType(object v)
		{
			return v == null;
		}
	}
	
	public class DATATYPE_INT8 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readInt8();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeInt8(Convert.ToSByte(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			SByte ret = 0;
			SByte.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= sbyte.MinValue && v1 <= sbyte.MaxValue;
		}
	}
	
	public class DATATYPE_INT16 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readInt16();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeInt16(Convert.ToInt16(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			Int16 ret = 0;
			Int16.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= Int16.MinValue && v1 <= Int16.MaxValue;
		}
	}
	
	public class DATATYPE_INT32 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readInt32();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeInt32(Convert.ToInt32(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			Int32 ret = 0;
			Int32.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= Int32.MinValue && v1 <= Int32.MaxValue;
		}
	}
	
	public class DATATYPE_INT64 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readInt64();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeInt64(Convert.ToInt64(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			Int64 ret = 0;
			Int64.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= Int64.MinValue && v1 <= Int64.MaxValue;
		}
	}
	
	public class DATATYPE_UINT8 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readUint8();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeUint8(Convert.ToByte(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			Byte ret = 0;
			Byte.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= Byte.MinValue && v1 <= Byte.MaxValue;
		}
	}
	
	public class DATATYPE_UINT16 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readUint16();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeUint16(Convert.ToUInt16(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			UInt16 ret = 0;
			UInt16.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= UInt16.MinValue && v1 <= UInt16.MaxValue;
		}
	}
	
	public class DATATYPE_UINT32 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readUint32();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeUint32(Convert.ToUInt32(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			UInt32 ret = 0;
			UInt32.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= UInt32.MinValue && v1 <= UInt32.MaxValue;
		}
	}
	
	public class DATATYPE_UINT64 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readUint64();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeUint64(Convert.ToUInt64(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			UInt64 ret = 0;
			UInt64.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if (!KBEMath.isNumeric (v))
				return false;

			decimal v1 = Convert.ToDecimal (v);
			return v1 >= UInt64.MinValue && v1 <= UInt64.MaxValue;
		}
	}
	
	public class DATATYPE_FLOAT : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readFloat();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeFloat((float)Convert.ToDouble(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			float ret = 0.0f;
			float.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if(v is float)
				return (float)v >= float.MinValue && (float)v <= float.MaxValue;
			else if(v is double)
				return (double)v >= float.MinValue && (double)v <= float.MaxValue;
			
			return false;
		}
	}
	
	public class DATATYPE_DOUBLE : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readDouble();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeDouble(Convert.ToDouble(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			double ret = 0.0;
			double.TryParse(v, out ret);
			return ret;
		}
		
		public override bool isSameType(object v)
		{
			if(v is float)
				return (float)v >= double.MinValue && (float)v <= double.MaxValue;
			else if(v is double)
				return (double)v >= double.MinValue && (double)v <= double.MaxValue;
			
			return false;
		}
	}
	
	public class DATATYPE_STRING : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readString();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeString(Convert.ToString(v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			return v;
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(string);
		}
	}
	
	public class DATATYPE_VECTOR2 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return new Vector2(stream.readFloat(), stream.readFloat());
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeFloat(((Vector2)v).x);
			stream.writeFloat(((Vector2)v).y);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new Vector2(0.0f, 0.0f);
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(Vector2);
		}
	}
	
	public class DATATYPE_VECTOR3 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return new Vector3(stream.readFloat(), stream.readFloat(), stream.readFloat());
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeFloat(((Vector3)v).x);
			stream.writeFloat(((Vector3)v).y);
			stream.writeFloat(((Vector3)v).z);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new Vector3(0.0f, 0.0f, 0.0f);
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(Vector3);
		}
	}
	
	public class DATATYPE_VECTOR4 : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return new Vector4(stream.readFloat(), stream.readFloat(), stream.readFloat(), stream.readFloat());
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeFloat(((Vector4)v).x);
			stream.writeFloat(((Vector4)v).y);
			stream.writeFloat(((Vector4)v).z);
			stream.writeFloat(((Vector4)v).w);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new Vector4(0.0f, 0.0f, 0.0f, 0.0f);
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(Vector4);
		}
	}
	
	public class DATATYPE_PYTHON : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readBlob();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeBlob((byte[])v);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new byte[0];
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(byte[]);
		}
	}
	
	public class DATATYPE_UNICODE : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return System.Text.Encoding.UTF8.GetString(stream.readBlob());
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeBlob(System.Text.Encoding.UTF8.GetBytes((string)v));
		}
		
		public override object parseDefaultValStr(string v)
		{
			return v;
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(string);
		}
	}
	
	public class DATATYPE_ENTITYCALL : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readBlob();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeBlob((byte[])v);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new byte[0];
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(byte[]);
		}
	}
	
	public class DATATYPE_BLOB : DATATYPE_BASE
	{
		public override object createFromStream(MemoryStream stream)
		{
			return stream.readBlob();
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeBlob((byte[])v);
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new byte[0];
		}
		
		public override bool isSameType(object v)
		{
			return v != null && v.GetType() == typeof(byte[]);
		}
	}
	
	public class DATATYPE_ARRAY : DATATYPE_BASE
	{
		public object vtype;
		
		public override void bind()
		{
			if(vtype.GetType ().BaseType.ToString() == "KBEngine.DATATYPE_BASE")
				((DATATYPE_BASE)vtype).bind();
			else
				if(EntityDef.id2datatypes.ContainsKey((UInt16)vtype))
					vtype = EntityDef.id2datatypes[(UInt16)vtype];
		}
		
		public override object createFromStream(MemoryStream stream)
		{
			UInt32 size = stream.readUint32();
			List<object> datas = new List<object>();
			
			while(size > 0)
			{
				size--;
		//		datas.Add(((DATATYPE_BASE)vtype).createFromStream(stream));
			};
			
			return datas;
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			stream.writeUint32((UInt32)((List<object>)v).Count);
			for(int i=0; i<((List<object>)v).Count; i++)
			{
		//		((DATATYPE_BASE)vtype).addToStream(stream, ((List<object>)v)[i]);
			}
		}
		
		public override object parseDefaultValStr(string v)
		{
			return new byte[0];
		}
		
		public override bool isSameType(object v)
		{
			if(vtype.GetType ().BaseType.ToString() != "KBEngine.DATATYPE_BASE")
			{
				Dbg.ERROR_MSG(string.Format("DATATYPE_ARRAY::isSameType: has not bind! baseType={0}", 
					vtype.GetType ().BaseType.ToString()));
				return false;
			}

			if(v == null || v.GetType() != typeof(List<object>))
			{
				return false;
			}
			
			for(int i=0; i<((List<object>)v).Count; i++)
			{
				if(!((DATATYPE_BASE)vtype).isSameType(((List<object>)v)[i]))
				{
					return false;
				}
			}
			
			return true;
		}
	}
	
	public class DATATYPE_FIXED_DICT : DATATYPE_BASE
	{
		public string implementedBy = "";
		public Dictionary<string, object> dicttype = new Dictionary<string, object>();
		
		public override void bind()
		{
			string[] keys = new string[dicttype.Keys.Count];
			dicttype.Keys.CopyTo(keys, 0);
			
			foreach(string itemkey in keys)
			{
				object type = dicttype[itemkey];
				
				if(type.GetType ().BaseType.ToString() == "KBEngine.DATATYPE_BASE")
					((DATATYPE_BASE)type).bind();
				else
					if(EntityDef.id2datatypes.ContainsKey((UInt16)type))
						dicttype[itemkey] = EntityDef.id2datatypes[(UInt16)type];
			}
		}
		
		public override object createFromStream(MemoryStream stream)
		{
			Dictionary<string, object> datas = new Dictionary<string, object>();
			foreach(string itemkey in dicttype.Keys)
			{
			//	datas[itemkey] = ((DATATYPE_BASE)dicttype[itemkey]).createFromStream(stream);
			}
			
			return datas;
		}
		
		public override void addToStream(Bundle stream, object v)
		{
			foreach(string itemkey in dicttype.Keys)
			{
		//		((DATATYPE_BASE)dicttype[itemkey]).addToStream(stream, ((Dictionary<string, object>)v)[itemkey]);
			}
		}
		
		public override object parseDefaultValStr(string v)
		{
			Dictionary<string, object> datas = new Dictionary<string, object>();
			foreach(string itemkey in dicttype.Keys)
			{
				datas[itemkey] = ((DATATYPE_BASE)dicttype[itemkey]).parseDefaultValStr("");
			}
			
			return datas;
		}
		
		public override bool isSameType(object v)
		{
			if(v == null || v.GetType() != typeof(Dictionary<string, object>))
				return false;
						
			foreach(KeyValuePair<string, object> item in dicttype)
			{
				object value;
				if(((Dictionary<string, object>)v).TryGetValue(item.Key, out value))
				{
					if(!((DATATYPE_BASE)item.Value).isSameType(value))
					{
						return false;
					}
				}
				else
				{
					return false;
				}
			}
			
			return true;
		}
	}
} 

File: kbe/res/sdk_templates/client/ue4/Source/KBEnginePlugins/Engine/DataTypes.cpp

#include "DataTypes.h"
#include "MemoryStream.h"
#include "EntityDef.h"
#include "KBDebug.h"
#include "Runtime/Core/Public/Misc/Variant.h"
#include "Bundle.h"

namespace KBEngine
{

KBVar* DATATYPE_INT8::createFromStream(MemoryStream& stream)
{
	int8 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_INT8::addToStream(Bundle& stream, KBVar& v)
{
	int8 val = v;
	stream << val;
}

KBVar* DATATYPE_INT8::parseDefaultValStr(const FString& v)
{
	int8 val = (int8)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_INT8::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Int8;
}

KBVar* DATATYPE_INT16::createFromStream(MemoryStream& stream)
{
	int16 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_INT16::addToStream(Bundle& stream, KBVar& v)
{
	int16 val = v;
	stream << val;
}

KBVar* DATATYPE_INT16::parseDefaultValStr(const FString& v)
{
	int16 val = (int16)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_INT16::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Int16;
}

KBVar* DATATYPE_INT32::createFromStream(MemoryStream& stream)
{
	int32 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_INT32::addToStream(Bundle& stream, KBVar& v)
{
	int32 val = v;
	stream << val;
}

KBVar* DATATYPE_INT32::parseDefaultValStr(const FString& v)
{
	int32 val = (int32)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_INT32::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Int32;
}

KBVar* DATATYPE_INT64::createFromStream(MemoryStream& stream)
{
	int64 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_INT64::addToStream(Bundle& stream, KBVar& v)
{
	int64 val = v;
	stream << val;
}

KBVar* DATATYPE_INT64::parseDefaultValStr(const FString& v)
{
	int64 val = (int64)FCString::Atoi64(*v);
	return new KBVar(val);
}

bool DATATYPE_INT64::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Int64;
}

KBVar* DATATYPE_UINT8::createFromStream(MemoryStream& stream)
{
	uint8 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_UINT8::addToStream(Bundle& stream, KBVar& v)
{
	uint8 val = v;
	stream << val;
}

KBVar* DATATYPE_UINT8::parseDefaultValStr(const FString& v)
{
	uint8 val = (uint8)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_UINT8::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::UInt8;
}

KBVar* DATATYPE_UINT16::createFromStream(MemoryStream& stream)
{
	uint16 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_UINT16::addToStream(Bundle& stream, KBVar& v)
{
	uint16 val = v;
	stream << val;
}

KBVar* DATATYPE_UINT16::parseDefaultValStr(const FString& v)
{
	uint16 val = (uint16)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_UINT16::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::UInt16;
}

KBVar* DATATYPE_UINT32::createFromStream(MemoryStream& stream)
{
	uint32 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_UINT32::addToStream(Bundle& stream, KBVar& v)
{
	uint32 val = v;
	stream << val;
}

KBVar* DATATYPE_UINT32::parseDefaultValStr(const FString& v)
{
	uint32 val = (uint32)FCString::Atoi(*v);
	return new KBVar(val);
}

bool DATATYPE_UINT32::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::UInt32;
}

KBVar* DATATYPE_UINT64::createFromStream(MemoryStream& stream)
{
	uint64 val = 0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_UINT64::addToStream(Bundle& stream, KBVar& v)
{
	uint64 val = v;
	stream << val;
}

KBVar* DATATYPE_UINT64::parseDefaultValStr(const FString& v)
{
	uint64 val = (uint64)FCString::Atoi64(*v);
	return new KBVar(val);
}

bool DATATYPE_UINT64::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::UInt64;
}

KBVar* DATATYPE_FLOAT::createFromStream(MemoryStream& stream)
{
	float val = 0.f;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_FLOAT::addToStream(Bundle& stream, KBVar& v)
{
	float val = v;
	stream << val;
}

KBVar* DATATYPE_FLOAT::parseDefaultValStr(const FString& v)
{
	float val = (float)FCString::Atof(*v);
	return new KBVar(val);
}

bool DATATYPE_FLOAT::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Float;
}

KBVar* DATATYPE_DOUBLE::createFromStream(MemoryStream& stream)
{
	double val = 0.0;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_DOUBLE::addToStream(Bundle& stream, KBVar& v)
{
	double val = v;
	stream << val;
}

KBVar* DATATYPE_DOUBLE::parseDefaultValStr(const FString& v)
{
	double val = (double)FCString::Atof(*v);
	return new KBVar(val);
}

bool DATATYPE_DOUBLE::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Double;
}

KBVar* DATATYPE_STRING::createFromStream(MemoryStream& stream)
{
	FString val;
	stream >> val;
	return new KBVar(val);
}

void DATATYPE_STRING::addToStream(Bundle& stream, KBVar& v)
{
	FString val = v;
	stream << val;
}

KBVar* DATATYPE_STRING::parseDefaultValStr(const FString& v)
{
	return new KBVar(v);
}

bool DATATYPE_STRING::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::String || v.GetType() == EKBVarTypes::Ansichar;
}

KBVar* DATATYPE_VECTOR2::createFromStream(MemoryStream& stream)
{
	FVector2D val;
	stream >> val.X >> val.Y;
	return new KBVar(val);
}

void DATATYPE_VECTOR2::addToStream(Bundle& stream, KBVar& v)
{
	FVector2D val = v;
	stream << val.X << val.Y;
}

KBVar* DATATYPE_VECTOR2::parseDefaultValStr(const FString& v)
{
	return new KBVar(FVector2D());
}

bool DATATYPE_VECTOR2::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Vector2d;
}

KBVar* DATATYPE_VECTOR3::createFromStream(MemoryStream& stream)
{
	FVector val;
	stream >> val.X >> val.Y >> val.Z;
	return new KBVar(val);
}

void DATATYPE_VECTOR3::addToStream(Bundle& stream, KBVar& v)
{
	FVector val = v;
	stream << val.X << val.Y << val.Z;
}

KBVar* DATATYPE_VECTOR3::parseDefaultValStr(const FString& v)
{
	return new KBVar(FVector());
}

bool DATATYPE_VECTOR3::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Vector;
}

KBVar* DATATYPE_VECTOR4::createFromStream(MemoryStream& stream)
{
	FVector4 val;
	stream >> val.X >> val.Y >> val.Z >> val.W;
	return new KBVar(val);
}

void DATATYPE_VECTOR4::addToStream(Bundle& stream, KBVar& v)
{
	FVector4 val = v;
	stream << val.X << val.Y << val.Z << val.W;
}

KBVar* DATATYPE_VECTOR4::parseDefaultValStr(const FString& v)
{
	return new KBVar(FVector4());
}

bool DATATYPE_VECTOR4::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::Vector4;
}

KBVar* DATATYPE_PYTHON::createFromStream(MemoryStream& stream)
{
	TArray<uint8> val;
	stream.readBlob(val);
	return new KBVar(val);
}

void DATATYPE_PYTHON::addToStream(Bundle& stream, KBVar& v)
{
	TArray<uint8> val = v;
	stream.appendBlob(val);
}

KBVar* DATATYPE_PYTHON::parseDefaultValStr(const FString& v)
{
	return new KBVar(TArray<uint8>());
}

bool DATATYPE_PYTHON::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::ByteArray;
}

KBVar* DATATYPE_UNICODE::createFromStream(MemoryStream& stream)
{
	FString val;
	stream.readUTF8String(val);
	return new KBVar(val);
}

void DATATYPE_UNICODE::addToStream(Bundle& stream, KBVar& v)
{
	FString val = v;
	stream.appendUTF8String(val);
}

KBVar* DATATYPE_UNICODE::parseDefaultValStr(const FString& v)
{
	return new KBVar(v);
}

bool DATATYPE_UNICODE::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::String || v.GetType() == EKBVarTypes::Widechar
		|| v.GetType() == EKBVarTypes::Ansichar;
}

KBVar* DATATYPE_ENTITYCALL::createFromStream(MemoryStream& stream)
{
	TArray<uint8> val;
	stream.readBlob(val);
	return new KBVar(val);
}

void DATATYPE_ENTITYCALL::addToStream(Bundle& stream, KBVar& v)
{
	TArray<uint8> val = v;
	stream.appendBlob(val);
}

KBVar* DATATYPE_ENTITYCALL::parseDefaultValStr(const FString& v)
{
	return new KBVar(TArray<uint8>());
}

bool DATATYPE_ENTITYCALL::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::ByteArray;
}

KBVar* DATATYPE_BLOB::createFromStream(MemoryStream& stream)
{
	TArray<uint8> val;
	stream.readBlob(val);
	return new KBVar(val);
}

void DATATYPE_BLOB::addToStream(Bundle& stream, KBVar& v)
{
	TArray<uint8> val = v;
	stream.appendBlob(val);
}

KBVar* DATATYPE_BLOB::parseDefaultValStr(const FString& v)
{
	return new KBVar(TArray<uint8>());
}

bool DATATYPE_BLOB::isSameType(KBVar& v)
{
	return v.GetType() == EKBVarTypes::ByteArray;
}

void DATATYPE_ARRAY::bind()
{
	if (tmpset_uitemtype == -1)
		vtype->bind();
	else
		if (EntityDef::id2datatypes.Contains(tmpset_uitemtype))
			vtype = EntityDef::id2datatypes[tmpset_uitemtype];
}

KBVar* DATATYPE_ARRAY::createFromStream(MemoryStream& stream)
{
	KBVar::KBVarArray val;

	uint32 size;
	stream >> size;

	while (size > 0)
	{
		size--;
		val.Add(*vtype->createFromStream(stream));
	};

	return new KBVar(val);
}

void DATATYPE_ARRAY::addToStream(Bundle& stream, KBVar& v)
{
	if (v.GetType() == EKBVarTypes::KBVarArray)
	{
		KBVar::KBVarArray val = v;
		uint32 size = val.Num();

		stream << size;

		for (uint32 i = 0; i<size; ++i)
		{
			vtype->addToStream(stream, val[i]);
		}
	}
	else
	{
		KBVar::KBVarBytes val = v;
		uint32 size = val.Num();

		stream << size;

		for (uint32 i = 0; i<size; ++i)
		{
			KBVar tmpv = val[i];
			vtype->addToStream(stream, tmpv);
		}
	}
}

KBVar* DATATYPE_ARRAY::parseDefaultValStr(const FString& v)
{
	return new KBVar(KBVar::KBVarArray());
}

bool DATATYPE_ARRAY::isSameType(KBVar& v)
{
	if (v.GetType() != EKBVarTypes::KBVarArray && v.GetType() != EKBVarTypes::ByteArray)
		return false;

	if (v.GetType() == EKBVarTypes::KBVarArray)
	{
		KBVar::KBVarArray val = v;

		for (uint32 i = 0; i < (uint32)val.Num(); ++i)
		{
			if (!vtype->isSameType(val[i]))
				return false;
		}
	}
	else
	{
		KBVar::KBVarBytes val = v;

		for (uint32 i = 0; i < (uint32)val.Num(); ++i)
		{
			KBVar tmpv = val[i];
			if (!vtype->isSameType(tmpv))
				return false;
		}
	}

	return true;
}

void DATATYPE_FIXED_DICT::bind()
{
	if (dicttype_map.Num() > 0)
	{
		for (auto& item : dicttype_map)
		{
			if (EntityDef::id2datatypes.Contains(item.Value))
				dicttype.Add(item.Key, EntityDef::id2datatypes[item.Value]);
		}

		dicttype_map.Empty();
	}
	else
	{
		for (auto& item : dicttype)
		{
			item.Value->bind();
		}
	}
}

KBVar* DATATYPE_FIXED_DICT::createFromStream(MemoryStream& stream)
{
	KBVar::KBVarMap val;

	for (auto& item : dicttype)
	{
		val.Add(item.Key, *item.Value->createFromStream(stream));
	}

	return new KBVar(val);
}

void DATATYPE_FIXED_DICT::addToStream(Bundle& stream, KBVar& v)
{
	KBVar::KBVarMap val = v;

	for (auto& item : dicttype)
	{
		item.Value->addToStream(stream, val[item.Key]);
	}
}

KBVar* DATATYPE_FIXED_DICT::parseDefaultValStr(const FString& v)
{
	KBVar::KBVarMap val;

	for (auto& item : dicttype)
	{
		val.Add(item.Key, *item.Value->parseDefaultValStr(FString(TEXT(""))));
	}

	return new KBVar(val);
}

bool DATATYPE_FIXED_DICT::isSameType(KBVar& v)
{
	if (v.GetType() != EKBVarTypes::KBVarMap)
		return false;

	KBVar::KBVarMap val = v;

	for (auto& item : dicttype)
	{
		if (!item.Value->isSameType(val[item.Key]))
			return false;
	}

	return true;
}

}